source/io/module_javaScriptInvoke.js

```diff
@@ -131,156 +138,123 @@ var assignJavaScriptInvoke;
         };
 
-        var createJavaScriptParametersArray = function (parametersPointer, parametersCount) {
-            var parameters = new Array(parametersCount);
-            for (var index = 0; index < parametersCount; index += 1) {
-                var typeName = getParameterTypeString(parametersPointer, index);
-                var parameterPointer = JavaScriptInvoke_GetParameterPointer(parametersPointer, index);
-                if (typeName === 'Array') {
-                    typeName += getArrayElementTypeString(parameterPointer);
+        var jsInvokePeeker = createJavaScriptInvokePeeker();
+        var peekValueRef = function (valueRef) {
+            return Module.eggShell.reflectOnValueRef(jsInvokePeeker, valueRef);
+        };
+
+        var createJavaScriptInvokePoker = function () {
+            var visitNumeric = function (valueRef, data) {
+                if (typeof data.userValue !== 'number') {
+                    mergeNewError(data.errorValueRef, data.functionName, ERRORS.kNITypeMismatchForReturnTypeInJavaScriptInvoke);
+                    return;
                 }
+                Module.eggShell.writeDouble(valueRef, data.userValue);
+            };
 
-                var parameterValue = undefined;
-                var readFunction = typeFunctions[typeName].reader;
-                if (readFunction === undefined) {
-                    throw new Error(' Unsupported type = ' + typeName + ' for parameter with index = ' + index);
-                } else {
-                    parameterValue = readFunction(parameterPointer);
+            return {
+                visitInt8: visitNumeric,
+                visitInt16: visitNumeric,
+                visitInt32: visitNumeric,
+                visitUInt8: visitNumeric,
+                visitUInt16: visitNumeric,
+                visitUInt32: visitNumeric,
+                visitSingle: visitNumeric,
+                visitDouble: visitNumeric,
+                visitEnum8: visitNumeric,
+                visitEnum16: visitNumeric,
+                visitEnum32: visitNumeric,
+
+                visitBoolean: function (valueRef, data) {
+                    if (typeof data.userValue !== 'boolean') {
+                        mergeNewError(data.errorValueRef, data.functionName, ERRORS.kNITypeMismatchForReturnTypeInJavaScriptInvoke);
+                        return;
+                    }
+                    Module.eggShell.writeDouble(valueRef, data.userValue ? 1 : 0);
+                },
+
+                visitString: function (valueRef, data) {
+                    if (typeof data.userValue !== 'string') {
+                        mergeNewError(data.errorValueRef, data.functionName, ERRORS.kNITypeMismatchForReturnTypeInJavaScriptInvoke);
+                        return;
+                    }
+                    Module.eggShell.writeString(valueRef, data.userValue);
+                },
+
+                visitArray: function (valueRef, data) {
+                    try {
```

Comment:
Update to check type of value to return better errors

Reply:
I was contemplating about passing error object to reflect APIs and missed this one earlier. Fixing in the next commit as discussed with Milan for now to check for type compatibility for typed arrays.